### PR TITLE
VP/MIN-302. RPC protocol total value.

### PR DESCRIPTION
**Pull request by dev-vadim**
**Description:**

Add cotroller RPC to return total amount of money in the protocol

**Changes in storage or API:**

Describe storage structure changes or API changes. Make sure to update the [API docs](https://minterestfinance.atlassian.net/wiki/spaces/MINTEREST/pages/134807557/Minterest+API).

**Checklist**

- [ ] Code is covered with tests
- [ ] Code is covered with benchmarks
- [ ] Fresh benchmark results included

### CODE REVIEW

File: pallets/controller/rpc/src/lib.rs [hunk 1]
```diff
@@ -15,6 +15,16 @@ use std::sync::Arc;
 #[rpc]
 /// Base trait for RPC interface of controller
 pub trait ControllerRpcApi<BlockHash, AccountId> {
+	/// Returns total amount of money currently held in the protocol in usd.
+	///
+	///  - `&self` :  Self reference
+	///  - `at` : Needed for runtime API use. Runtime API must always be called at a specific block.
+	///
+	/// Return:
+	/// - amount: account total collateral converted to usd.
```
Comment (lut1k) [inline pallets/controller/rpc/src/lib.rs:24]: Something went wrong

File: pallets/controller/src/lib.rs [hunk 2]
```diff
@@ -829,6 +842,17 @@ impl<T: Config> Pallet<T> {
 		Ok(current_block_number - accrual_block_number_previous)
 	}
 
+	/// Returns pool parameters calculated for a current block.
+	/// - `pool_id`: CurrencyId to calculate parameters for.
+	///
+	/// returns `current_block_number - accrual_block_number_previous`
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:848]: This function returns pool parameters, not block delta

File: runtime/src/tests/mod.rs [hunk 3]
```diff
@@ -494,6 +504,10 @@ fn get_all_freshest_prices() -> Vec<(CurrencyId, Option<Price>)> {
 	<Runtime as PricesRuntimeApi<Block>>::get_all_freshest_prices()
 }
 
+fn lock_price(currency_id: CurrencyId) -> DispatchResultWithPostInfo {
```
Comment (lut1k) [inline runtime/src/tests/mod.rs:507]: function is never used: `lock_price`

File: pallets/controller/src/lib.rs [hunk 4]
```diff
@@ -585,6 +577,35 @@ impl<T: Config> Pallet<T> {
...
+					.checked_mul(&oracle_price)
+					.map(|val| val.into_inner())
+					.ok_or(Error::<T>::NumOverflow)?;
+
+				Ok(current_value
+					.checked_add(pool_total_value_usd)
+					.ok_or(Error::<T>::NumOverflow)?)
+			})?;
+		Ok(total_value)
+	}
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:607]:
```suggestion
	/// Calculates total amount of money currently held in the protocol in usd.
	/// Total value is calculated as: sum(total_issuance_n * exchange_rate_n * oracle_price_n),
	/// where: `total_issuance_n` - total number of wrapped tokens in the n pool;
	/// 		`exchange_rate_n` - exchange rate in the n pool;
	/// 		`oracle_price_n` - oracle price for the n pool.
	pub fn get_protocol_total_value() -> BalanceResult {
		let total_value = CurrencyId::get_enabled_tokens_in_protocol(UnderlyingAsset)
			.iter()
			.filter(|&underlying_id| T::LiquidityPoolsManager::pool_exists(underlying_id))
			.try_fold(Balance::zero(), |current_value, &pool_id| -> BalanceResult {
				let pool_data = Self::calculate_current_pool_data(pool_id)?;
				let wrapped_id = pool_id.wrapped_asset().ok_or(Error::<T>::NotValidUnderlyingAssetId)?;
				let wrapped_balance = T::MultiCurrency::total_issuance(wrapped_id);

				let current_exchange_rate = <LiquidityPools<T>>::get_exchange_rate_by_interest_params(
					pool_id,
					pool_data.total_protocol_interest,
					pool_data.total_borrowed,
				)?;
				let oracle_price = T::PriceSource::get_underlying_price(pool_id).ok_or(Error::<T>::InvalidFeedPrice)?;

				let pool_total_value_usd = Rate::from_inner(wrapped_balance)
					.checked_mul(&current_exchange_rate)
					.and_then(|v| v.checked_mul(&oracle_price))
					.map(|x| x.into_inner())
					.ok_or(Error::<T>::BalanceOverflow)?;

				Ok(current_value
					.checked_add(pool_total_value_usd)
					.ok_or(Error::<T>::NumOverflow)?)
			})?;
		Ok(total_value)
	}
```

Comment (lut1k) [reply pallets/controller/src/lib.rs:607]: In my opinion, it is better to calculate through the exchange rate)

Comment (dev-vadim) [reply pallets/controller/src/lib.rs]: Done

File: runtime/src/tests/mod.rs (hunk 3 above)
Comment (dev-vadim) [reply runtime/src/tests/mod.rs:507]: Fixed

File: pallets/controller/src/lib.rs [hunk 5]
```diff
@@ -829,6 +842,17 @@ impl<T: Config> Pallet<T> {
 		Ok(current_block_number - accrual_block_number_previous)
 	}
 
+	/// Returns pool parameters calculated for a current block.
+	/// - `pool_id`: CurrencyId to calculate parameters for.
+	///
+	/// returns `current_block_number - accrual_block_number_previous`
+	pub fn calculate_current_pool_data(pool_id: CurrencyId) -> result::Result<Pool, DispatchError> {
```
[the remark this answers is not in the document]
Comment (dev-vadim) [reply pallets/controller/src/lib.rs:855]: Done

File: pallets/controller/src/lib.rs (hunk 2 above)
Comment (dev-vadim) [reply pallets/controller/src/lib.rs]: Fixed

File: pallets/controller/rpc/src/lib.rs (hunk 1 above)
Comment (dev-vadim) [reply pallets/controller/rpc/src/lib.rs]: Fixed

**Review by lut1k — APPROVED** [review]

**Review by Apollo50 — APPROVED** [review]
looks good for me!

**Review by kurotych — APPROVED** [review]
Great :+1:

**Pull request merged**